File: src/utils/spikeAttributes.js
const finiteNumber = (value) => {
  if (typeof value === 'boolean') return null;
  const number = Number(value);
  return Number.isFinite(number) ? number : null;
};

const clusterIdOf = (cluster, fallback) => cluster?.clusterId ?? cluster?.id ?? fallback;
const safeId = (value) => String(value).replace(/[^a-zA-Z0-9_.-]+/g, '_').replace(/^_+|_+$/g, '') || 'value';
const labelOf = (value) => String(value)
  .replace(/([A-Z])/g, ' $1')
  .replace(/_/g, ' ')
  .trim()
  .replace(/\b\w/g, (character) => character.toUpperCase());
const units = { channel: 'channel_id', depth: 'um', probeDepth: 'um', ypos: 'um', yPosition: 'um' };
const reserved = new Set([
  'time', 'x', 'y', 'spikeIndex', 'spike_index', 'clusterId', 'cluster_id',
  'features', 'template_features', 'templateFeatures', 'pointIndex',
]);

const numericVector = (value) => {
  if (!Array.isArray(value) || value.length < 2 || value.length > 32) return null;
  const vector = value.map(finiteNumber);
  return vector.every((number) => number !== null) ? vector : null;
};

const scalarDefinition = (id, field, provenance, label = labelOf(field), unit = null) => ({
  id,
  label,
  shape: 'scalar',
  dimensions: [{ id: 'value', label, unit }],
  provenance,
  kind: 'scalar',
  field,
});

const vectorDefinition = (id, field, provenance, labels, kind = 'vector') => ({
  id,
  label: labelOf(field),
  shape: 'two_dimensional',
  dimensions: [
    { id: 'x', label: labels?.[0] || `${labelOf(field)} 1`, unit: null },
    { id: 'y', label: labels?.[1] || `${labelOf(field)} 2`, unit: null },
  ],
  provenance,
  kind,
  field,
});

const sourceEntries = ({ clusterData, clusteringResults }) => {
  if (Array.isArray(clusteringResults?.fullData) && clusteringResults.fullData.length) {
    return clusteringResults.fullData.map((spikes, index) => ({
      clusterId: clusterIdOf(clusteringResults.clusters?.[index], index),
      spikes: (spikes || []).map((spike, pointIndex) => ({ ...spike, pointIndex })),
    }));
  }
  return (clusterData?.clusters || []).map((cluster, index) => {
    const points = cluster.points || [];
    const times = cluster.spikeTimes || cluster.spike_times || [];
    const channels = cluster.spikeChannels || [];
    const amplitudes = cluster.spikeAmplitudes || [];
    const count = Math.max(points.length, times.length, channels.length, amplitudes.length);
    return {
      clusterId: clusterIdOf(cluster, index),
      spikes: Array.from({ length: count }, (_, pointIndex) => ({
        x: points[pointIndex]?.[0],
        y: points[pointIndex]?.[1],
        time: times[pointIndex],
        channel: channels[pointIndex] ?? cluster.channelId,
        amplitude: amplitudes[pointIndex],
        spikeIndex: cluster.spikeIndices?.[pointIndex] ?? pointIndex,
        pointIndex,
      })),
    };
  });
};

const discoverDefinitions = (entries) => {
  const definitions = new Map();
  entries.forEach((entry) => {
    const spikes = entry.spikes.length <= 256
      ? entry.spikes
      : Array.from({ length: 256 }, (_, index) => entry.spikes[
          Math.round(index * (entry.spikes.length - 1) / 255)
        ]);
    spikes.forEach((spike) => {
      if (finiteNumber(spike.x) !== null && finiteNumber(spike.y) !== null) {
        definitions.set('embedding', definitions.get('embedding') || vectorDefinition(
          'embedding',
          'PCA embedding',
          { source: 'retained_embedding', fields: ['x', 'y'] },
          ['PC 1', 'PC 2'],
          'embedding'
        ));
      }
      Object.entries(spike).forEach(([key, value]) => {
        if (reserved.has(key)) return;
        const safeKey = safeId(key);
        if (finiteNumber(value) !== null) {
          const id = `metadata:${safeKey}`;
          definitions.set(id, definitions.get(id) || scalarDefinition(
            id, key, { source: 'spike_metadata', field: key }, labelOf(key), units[key]
          ));
        } else if (numericVector(value)) {
          const id = `vector:${safeKey}`;
          definitions.set(id, definitions.get(id) || vectorDefinition(
            id, key, { source: 'spike_metadata_vector', field: key, indices: [0, 1] }
          ));
        }
      });
      if (spike.features && !Array.isArray(spike.features) && typeof spike.features === 'object') {
        Object.entries(spike.features).forEach(([key, value]) => {
          if (finiteNumber(value) === null) return;
          const id = `feature:${safeId(key)}`;
          definitions.set(id, definitions.get(id) || scalarDefinition(
            id, key, { source: 'retained_feature_map', field: key }
          ));
        });
      } else if (numericVector(spike.features)) {
        definitions.set('features', definitions.get('features') || vectorDefinition(
          'features',
          'Feature vector',
          { source: 'retained_feature_vector', field: 'features', indices: [0, 1] },
          ['Feature 1', 'Feature 2'],
          'features_vector'
        ));
      }
      const templateFeatures = spike.template_features ?? spike.templateFeatures;
      if (numericVector(templateFeatures)) {
        definitions.set('template_features', definitions.get('template_features') || vectorDefinition(
          'template_features',
          'Template features',
          { source: 'retained_template_features', indices: [0, 1] },
          ['Template feature 1', 'Template feature 2'],
          'template_features'
        ));
      }
    });
  });
  const preferred = ['embedding', 'template_features', 'features', 'metadata:amplitude', 'metadata:channel'];
  const ids = [
    ...preferred.filter((id) => definitions.has(id)),
    ...[...definitions.keys()].filter((id) => !preferred.includes(id)).sort(),
  ];
  return ids.slice(0, 128).map((id) => definitions.get(id));
};

const valuesFor = (spike, definition) => {
  if (definition.kind === 'embedding') {
    const x = finiteNumber(spike.x);
    const y = finiteNumber(spike.y);
    return x === null || y === null ? null : { x, y };
  }
  if (definition.kind === 'scalar') {
    const raw = definition.provenance.source === 'retained_feature_map'
      ? spike.features?.[definition.provenance.field]
      : spike[definition.field];
    const value = finiteNumber(raw);
    return value === null ? null : { value };
  }
  const vector = definition.kind === 'features_vector'
    ? numericVector(spike.features)
    : definition.kind === 'template_features'
    ? numericVector(spike.template_features ?? spike.templateFeatures)
    : numericVector(spike[definition.field]);
  return vector ? { x: vector[0], y: vector[1] } : null;
};

const publicDefinition = (definition) => definition ? ({
  id: definition.id,
  label: definition.label,
  shape: definition.shape,
  dimensions: definition.dimensions,
  provenance: definition.provenance,
}) : null;

const evenlySample = (points, maximum) => {
  if (points.length <= maximum) return points;
  return Array.from({ length: maximum }, (_, index) => points[
    Math.round(index * (points.length - 1) / (maximum - 1))
  ]);
};

export const buildLocalSpikeAttributePayload = ({
  clusterIds = [],
  clusterData = null,
  clusteringResults = null,
  attributeId = '',
  sampleRateHz = 30000,
  maxSpikesPerCluster = 5000,
} = {}) => {
  const rate = Math.max(finiteNumber(sampleRateHz) ?? 30000, 1);
  const allEntries = sourceEntries({ clusterData, clusteringResults });
  const selectedKeys = new Set(clusterIds.map(String));
  const entries = allEntries.filter((entry) => selectedKeys.has(String(entry.clusterId)));
  const definitions = discoverDefinitions(entries);
  const definition = definitions.find((candidate) => candidate.id === attributeId) || definitions[0] || null;
  const series = clusterIds.map((clusterId) => {
    const entry = entries.find((candidate) => String(candidate.clusterId) === String(clusterId))
      || { spikes: [] };
    const candidates = definition ? entry.spikes.map((spike, fallbackIndex) => {
      const values = valuesFor(spike, definition);
      const timeSamples = finiteNumber(spike.time ?? spike.timeSamples);
      if (!values || (definition.shape === 'scalar' && timeSamples === null)) return null;
      const pointIndex = finiteNumber(spike.pointIndex) ?? fallbackIndex;
      const spikeIndex = finiteNumber(spike.spikeIndex ?? spike.spike_index) ?? pointIndex;
      const channel = finiteNumber(spike.channel);
      return {
        spikeId: `${clusterId}:${Math.round(spikeIndex)}`,
        clusterId,
        pointIndex: Math.round(pointIndex),
        spikeIndex: Math.round(spikeIndex),
        timeSamples,
        timeSeconds: timeSamples === null ? null : timeSamples / rate,
        channel: channel === null ? null : Math.round(channel),
        values,
      };
    }).filter(Boolean).sort((left, right) => (
      (left.timeSamples ?? left.pointIndex) - (right.timeSamples ?? right.pointIndex)
    )) : [];
    const points = evenlySample(candidates, Math.max(10, maxSpikesPerCluster));
    return {
      clusterId,
      points,
      totalPoints: candidates.length,
      returnedPoints: points.length,
    };
  });
  return {
    clusterIds,
    sampleRateHz: rate,
    attributeDefinitions: definitions.map(publicDefinition),
    selectedAttributeId: definition?.id || null,
    attributeDefinition: publicDefinition(definition),
    series,
  };
};

export const getSpikeAttributeSelection = (plotEvent) => {
  const points = (plotEvent?.points || [])
    .map((point) => point.customdata)
    .filter((point) => point?.spikeId);
  return [...new Map(points.map((point) => [point.spikeId, point])).values()];
};

File: processing/spike_attributes.py
"""Schema discovery and extraction for generic per-spike attribute views."""

import re

import numpy as np

from processing.cluster_diagnostics import normalize_cluster_ids


RESERVED_FIELDS = {
    'time', 'x', 'y', 'spikeIndex', 'spike_index', 'clusterId', 'cluster_id',
    'features', 'template_features', 'templateFeatures',
}
FIELD_UNITS = {
    'channel': 'channel_id',
    'depth': 'um',
    'probeDepth': 'um',
    'ypos': 'um',
    'yPosition': 'um',
}


def _finite_number(value):
    if isinstance(value, (bool, np.bool_)):
        return None
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return number if np.isfinite(number) else None


def _safe_id(value):
    return re.sub(r'[^a-zA-Z0-9_.-]+', '_', str(value)).strip('_') or 'value'


def _label(value):
    spaced = re.sub(r'(?<!^)(?=[A-Z])', ' ', str(value)).replace('_', ' ')
    return spaced.strip().title()


def _sample_for_schema(spikes, maximum=256):
    spikes = list(spikes or [])
    if len(spikes) <= maximum:
        return spikes
    indices = np.linspace(0, len(spikes) - 1, num=maximum, dtype=np.int64)
    return [spikes[index] for index in indices]


def _numeric_vector(value, minimum=2, maximum=32):
    if not isinstance(value, (list, tuple, np.ndarray)):
        return None
    try:
        vector = np.asarray(value, dtype=np.float64).reshape(-1)
    except (TypeError, ValueError):
        return None
    if vector.size < minimum or vector.size > maximum or not np.all(np.isfinite(vector)):
        return None
    return vector


def _scalar_definition(attribute_id, field, provenance, label=None, unit=None):
    return {
        'id': attribute_id,
        'label': label or _label(field),
        'shape': 'scalar',
        'dimensions': [{'id': 'value', 'label': label or _label(field), 'unit': unit}],
        'provenance': provenance,
        '_kind': 'scalar',
        '_field': field,
    }


def _vector_definition(attribute_id, field, provenance, labels=None, kind='vector'):
    dimension_labels = labels or (f'{_label(field)} 1', f'{_label(field)} 2')
    return {
        'id': attribute_id,
        'label': _label(field),
        'shape': 'two_dimensional',
        'dimensions': [
            {'id': 'x', 'label': dimension_labels[0], 'unit': None},
            {'id': 'y', 'label': dimension_labels[1], 'unit': None},
        ],
        'provenance': provenance,
        '_kind': kind,
        '_field': field,
    }


def _discover_specs(clustering_results, cluster_ids):
    specs = {}
    for cluster_id in cluster_ids:
        for spike in _sample_for_schema(clustering_results[cluster_id]):
            if not isinstance(spike, dict):
                continue
            if _finite_number(spike.get('x')) is not None and _finite_number(spike.get('y')) is not None:
                specs.setdefault('embedding', _vector_definition(
                    'embedding',
                    'PCA embedding',
                    {'source': 'retained_embedding', 'fields': ['x', 'y']},
                    labels=('PC 1', 'PC 2'),
                    kind='embedding',
                ))

            for key, value in spike.items():
                if key in RESERVED_FIELDS:
                    continue
                if _finite_number(value) is not None:
                    safe_key = _safe_id(key)
                    specs.setdefault(f'metadata:{safe_key}', _scalar_definition(
                        f'metadata:{safe_key}',
                        key,
                        {'source': 'spike_metadata', 'field': key},
                        unit=FIELD_UNITS.get(key),
                    ))
                    continue
                vector = _numeric_vector(value)
                if vector is not None:
                    safe_key = _safe_id(key)
                    specs.setdefault(f'vector:{safe_key}', _vector_definition(
                        f'vector:{safe_key}',
                        key,
                        {'source': 'spike_metadata_vector', 'field': key, 'indices': [0, 1]},
                    ))

            features = spike.get('features')
            if isinstance(features, dict):
                for key, value in features.items():
                    if _finite_number(value) is None:
                        continue
                    safe_key = _safe_id(key)
                    specs.setdefault(f'feature:{safe_key}', _scalar_definition(
                        f'feature:{safe_key}',
                        key,
                        {'source': 'retained_feature_map', 'field': str(key)},
                    ))
            elif _numeric_vector(features) is not None:
                specs.setdefault('features', _vector_definition(
                    'features',
                    'Feature vector',
                    {'source': 'retained_feature_vector', 'field': 'features', 'indices': [0, 1]},
                    labels=('Feature 1', 'Feature 2'),
                    kind='features_vector',
                ))

            template_features = spike.get('template_features', spike.get('templateFeatures'))
            if _numeric_vector(template_features) is not None:
                specs.setdefault('template_features', _vector_definition(
                    'template_features',
                    'Template features',
                    {'source': 'retained_template_features', 'indices': [0, 1]},
                    labels=('Template feature 1', 'Template feature 2'),
                    kind='template_features',
                ))

    preferred = ['embedding', 'template_features', 'features', 'metadata:amplitude', 'metadata:channel']
    ordered_ids = [attribute_id for attribute_id in preferred if attribute_id in specs]
    ordered_ids.extend(sorted(set(specs) - set(ordered_ids)))
    return [specs[attribute_id] for attribute_id in ordered_ids[:128]]


def _extract_values(spike, spec):
    kind = spec['_kind']
    if kind == 'embedding':
        values = (_finite_number(spike.get('x')), _finite_number(spike.get('y')))
    elif kind == 'scalar':
        source = spec['provenance']['source']
        if source == 'retained_feature_map':
            features = spike.get('features')
            value = features.get(spec['provenance']['field']) if isinstance(features, dict) else None
        else:
            value = spike.get(spec['_field'])
        number = _finite_number(value)
        return {'value': number} if number is not None else None
    elif kind == 'features_vector':
        vector = _numeric_vector(spike.get('features'))
        values = (vector[0], vector[1]) if vector is not None else (None, None)
    elif kind == 'template_features':
        vector = _numeric_vector(spike.get('template_features', spike.get('templateFeatures')))
        values = (vector[0], vector[1]) if vector is not None else (None, None)
    else:
        vector = _numeric_vector(spike.get(spec['_field']))
        values = (vector[0], vector[1]) if vector is not None else (None, None)
    if any(value is None for value in values):
        return None
    return {'x': float(values[0]), 'y': float(values[1])}


def _public_definition(spec):
    return {key: value for key, value in spec.items() if not key.startswith('_')}


def extract_spike_attribute(
    clustering_results,
    cluster_ids,
    attribute_id=None,
    sample_rate_hz=30000.0,
    max_spikes_per_cluster=5000,
):
    """Discover compatible attributes and extract one bounded typed payload."""
    sample_rate_hz = max(float(sample_rate_hz), 1.0)
    maximum = min(max(int(max_spikes_per_cluster), 10), 20000)
    cluster_ids = normalize_cluster_ids(clustering_results, cluster_ids, limit=12)
    specs = _discover_specs(clustering_results, cluster_ids)
    spec = next((item for item in specs if item['id'] == attribute_id), None)
    if spec is None:
        spec = specs[0] if specs else None

    series = []
    for cluster_id in cluster_ids:
        candidates = []
        if spec is not None:
            for point_index, spike in enumerate(clustering_results[cluster_id] or []):
                values = _extract_values(spike, spec)
                time_samples = _finite_number(spike.get('time')) if isinstance(spike, dict) else None
                if values is None or (spec['shape'] == 'scalar' and time_samples is None):
                    continue
                raw_spike_index = spike.get('spikeIndex', spike.get('spike_index', point_index))
                spike_index = _finite_number(raw_spike_index)
                spike_index = int(round(spike_index)) if spike_index is not None else point_index
                channel = _finite_number(spike.get('channel'))
                candidates.append({
                    'spikeId': f'{cluster_id}:{spike_index}',
                    'clusterId': cluster_id,
                    'pointIndex': point_index,
                    'spikeIndex': spike_index,
                    'timeSamples': time_samples,
                    'timeSeconds': time_samples / sample_rate_hz if time_samples is not None else None,
                    'channel': int(round(channel)) if channel is not None else None,
                    'values': values,
                })
        candidates.sort(key=lambda point: (
            point['timeSamples'] if point['timeSamples'] is not None else point['pointIndex']
        ))
        total_points = len(candidates)
        if len(candidates) > maximum:
            indices = np.linspace(0, len(candidates) - 1, num=maximum, dtype=np.int64)
            candidates = [candidates[index] for index in indices]
        series.append({
            'clusterId': cluster_id,
            'points': candidates,
            'totalPoints': total_points,
            'returnedPoints': len(candidates),
        })

    return {
        'clusterIds': cluster_ids,
        'sampleRateHz': sample_rate_hz,
        'attributeDefinitions': [_public_definition(item) for item in specs],
        'selectedAttributeId': spec['id'] if spec is not None else None,
        'attributeDefinition': _public_definition(spec) if spec is not None else None,
        'series': series,
    }
